### PR TITLE
Bug 1862289 - Translations UI Integration Low Data Translate

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1862289

### CODE REVIEW

File: fenix/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -2377,6 +2377,12 @@
     <string name="translation_error_language_not_supported_learn_more">Learn more</string>
     <!-- Snackbar title shown if the user closes the Translation Request dialogue and a translation is in progress. -->
     <string name="translation_in_progress_snackbar">Translating…</string>
+    <!-- Title for the data saving mode warning dialog used in translation request dialog.
+    This dialog will be presented when the user attempts to perform
+    a translation without the necessary language files downloaded first when Android's data saver mode is enabled and the user is not using WiFi.
+    The first parameter is the size in kilobytes or megabytes of the language file.-->
```
Comment (gabrielluong) [inline fenix/app/src/main/res/values/strings.xml:2383]:
```suggestion
    The first parameter is the size in kilobytes or megabytes of the language file. -->
```

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt [hunk 2]
```diff
@@ -139,6 +143,18 @@ sealed class TranslationsDialogAction : Action {
      * Updates the dialog title if the page was translated.
      */
     data class UpdateTranslatedPageTitle(val title: String) : TranslationsDialogAction()
+
+    /**
+     * Updates the translation download size file.
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:148]:
```suggestion
     * Updates the translation download file size.
```

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt [hunk 3]
```diff
@@ -139,6 +143,18 @@ sealed class TranslationsDialogAction : Action {
...
+     */
+    data class UpdateDownloadTranslationDownloadSize(val translationDownloadSize: TranslationDownloadSize? = null) :
+        TranslationsDialogAction()
+
+    /**
+     * Fetch the translation download size file.
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:154]:
```suggestion
     * Fetch the translation download file size.
```

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt [hunk 4]
```diff
@@ -225,21 +257,58 @@ class TranslationsDialogFragment : BottomSheetDialogFragment() {
                     }
                     dismiss()
                 },
-                onFromSelected = {
+                onFromSelected = { fromLanguageSelected ->
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt:260]:
```suggestion
                onFromSelected = { fromLanguage ->
```
Sometimes it's just better to just name it to what the named parameter is expecting.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt [hunk 5]
```diff
@@ -225,21 +257,58 @@ class TranslationsDialogFragment : BottomSheetDialogFragment() {
...
+
                     translationsDialogStore.dispatch(
                         TranslationsDialogAction.UpdateFromSelectedLanguage(
-                            it,
+                            fromLanguageSelected,
                         ),
                     )
                 },
-                onToSelected = {
+                onToSelected = { toLanguageSelected ->
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt:276]:
```suggestion
                onToSelected = { toLanguage ->
```

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt [hunk 6]
```diff
@@ -58,6 +62,9 @@ class TranslationsDialogFragment : BottomSheetDialogFragment() {
     private val translationDialogBinding = ViewBoundFeatureWrapper<TranslationsDialogBinding>()
     private lateinit var translationsDialogStore: TranslationsDialogStore
     private var isTranslationInProgress: Boolean? = null
+    private val downloadLanguagesFeature =
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt:65]: Could we organize these variables a bit better? Just moving `ViewBoundFeatureWrapper` together would be a good start.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt [hunk 7]
```diff
@@ -77,29 +77,35 @@ class TranslationsDialogBinding(
...
-
-                fromSelected?.let {
-                    translationsDialogStore.dispatch(
-                        TranslationsDialogAction.UpdateFromSelectedLanguage(
-                            fromSelected,
-                        ),
-                    )
+                // Dispatch initialFrom Language only the first time when it is null.
+                if (translationsDialogStore.state.initialFrom == null) {
+                    fromSelected?.let {
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:87]:
```suggestion
                if (fromSelected != null && translationsDialogStore.state.initialFrom == null) {
```
Maybe we should skip a subsequent check here if we just did it all in the if statement conditional

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt [hunk 8]
```diff
@@ -77,29 +77,35 @@ class TranslationsDialogBinding(
...
                     )
-                toSelected?.let {
-                    translationsDialogStore.dispatch(
-                        TranslationsDialogAction.UpdateToSelectedLanguage(
-                            toSelected,
-                        ),
-                    )
+                // Dispatch initialTo Language only the first time when it is null.
+                if (translationsDialogStore.state.initialTo == null) {
+                    toSelected?.let {
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:102]: Same as above

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt [hunk 9]
```diff
@@ -130,6 +136,12 @@ class TranslationsDialogBinding(
                         TranslationsDialogAction.UpdateTranslationError(sessionTranslationsState.translationError),
                     )
                 }
+
+                sessionTranslationsState.translationDownloadSize?.let {
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:140]: I added this check in reducer. ToSelected and fromSelected can be changed by the user from DropDowns, and the code from TranslationsDialogBinding doesn't get called.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt [hunk 10]
```diff
@@ -153,11 +165,13 @@ class TranslationsDialogBinding(
             ),
         )
 
-        translationsDialogStore.dispatch(
-            TranslationsDialogAction.UpdateTranslated(
-                true,
-            ),
-        )
+        if (!translationsDialogStore.state.isTranslated) {
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:168]: When I called  TranslationsAction.FetchTranslationDownloadSizeAction a new translationsState will come to the TranslationsDialogBinding and this line 
translationsDialogStore.dispatch(
                TranslationsDialogAction.UpdateTranslated(
                    true,
                ),
            )
will be called again making the positiveButton disabled. 
TranslationsDialogAction.UpdateTranslated -> {
                state.copy(
                    isTranslated = action.isTranslated,
                    positiveButtonType = PositiveButtonType.Disabled,
                )
            }
and I don't what this.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/preferences/downloadlanguages/DownloadLanguageFileDialog.kt [hunk 11]
```diff
@@ -163,24 +177,45 @@ private fun DownloadLanguageFileDialogCheckbox(
...
+ */
+sealed class DownloadLanguageFileDialogType {
+
+    /**
+     * All language files need to be downloaded.
+     */
+    data object AllLanguages : DownloadLanguageFileDialogType()
+
+    /**
+     * Only one default file need to be downloaded.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/preferences/downloadlanguages/DownloadLanguageFileDialog.kt:201]: Yes, done

File: fenix/app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -2377,6 +2377,12 @@
     <string name="translation_error_language_not_supported_learn_more">Learn more</string>
     <!-- Snackbar title shown if the user closes the Translation Request dialogue and a translation is in progress. -->
     <string name="translation_in_progress_snackbar">Translating…</string>
+    <!-- Title for the data saving mode warning dialog used in translation request dialog.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/res/values/strings.xml:2380]: Done

File: fenix/app/src/main/res/values/strings.xml (hunk 1 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/res/values/strings.xml:2383]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt (hunk 2 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:148]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt (hunk 3 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:154]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt (hunk 4 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt:260]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt (hunk 5 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt:276]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt (hunk 6 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt:65]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt (hunk 7 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:87]: DOne

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt (hunk 8 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:102]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt [hunk 13]
```diff
@@ -77,29 +77,35 @@ class TranslationsDialogBinding(
...
                         translateFromLanguages,
                     )
-
-                fromSelected?.let {
-                    translationsDialogStore.dispatch(
-                        TranslationsDialogAction.UpdateFromSelectedLanguage(
-                            fromSelected,
-                        ),
-                    )
+                // Dispatch initialFrom Language only the first time when it is null.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogBinding.kt:85]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/preferences/downloadlanguages/DownloadLanguageFileDialog.kt [hunk 14]
```diff
@@ -163,24 +177,45 @@ private fun DownloadLanguageFileDialogCheckbox(
...
+     */
+    data object Default : DownloadLanguageFileDialogType()
+
+    /**
+     * When the user presses the translate button, the site needs to be translated.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/preferences/downloadlanguages/DownloadLanguageFileDialog.kt:206]: Done

**Pull request merged**